### PR TITLE
Store removed objects rather than copy

### PR DESCRIPTION
**Pull request by KingOfBrian**

### CODE REVIEW

File: RZAssemblage/RZArrayAssemblage.m [hunk 1]
```diff
@@ -77,8 +77,9 @@ - (void)removeObjectFromChildrenAtIndex:(NSUInteger)index
 {
     RZAssemblageLog(@"%p:Remove %@ at %zd", self, [self objectInChildrenAtIndex:index],  index);
     [self openBatchUpdate];
+    id object = [self.childrenStorage objectAtIndex:index];
```
Comment (jvisenti) [inline RZAssemblage/RZArrayAssemblage.m:80]: You should implement

``` obj-c
- (id)objectAtIndexedSubscript:(NSUInteger)idx;
- (void)setObject:(id)obj atIndexedSubscript:(NSUInteger)idx;
```

So you can do things like `self.childrenStorage[index]`

File: RZAssemblage/RZAssemblageChangeSet.h [hunk 2]
```diff
@@ -18,11 +18,12 @@ typedef NSIndexPath *(^RZAssemblageChangeSetIndexPathTransform)(NSIndexPath *ind
...
 
-@property (strong, nonatomic) RZAssemblage *snapshot;
-
 @property (strong, nonatomic, readonly) NSArray *insertedIndexPaths;
 @property (strong, nonatomic, readonly) NSArray *updatedIndexPaths;
 @property (strong, nonatomic, readonly) NSArray *removedIndexPaths;
+
+- (id)removedObjectAtIndexPath:(NSIndexPath *)indexPath;
+
 @property (strong, nonatomic, readonly) NSDictionary *moveFromToIndexPaths;
```
Comment (jvisenti) [inline RZAssemblage/RZAssemblageChangeSet.h:27]: I like when methods come _after_ all the properties (style)

File: RZAssemblage/RZAssemblageChangeSet.m [hunk 3]
```diff
@@ -74,6 +74,14 @@ - (NSDictionary *)moveFromToIndexPaths
     return [self.moveFromToIndexPathMap copy];
 }
 
+- (id)removedObjectAtIndexPath:(NSIndexPath *)indexPath;
+{
+    NSParameterAssert(indexPath);
+    id object = [self.removedObjectsByIndexPath objectForKey:indexPath];
```
Comment (jvisenti) [inline RZAssemblage/RZAssemblageChangeSet.m:80]: `self.removedObjectsByIndexPath[indexPath]` for consistency

File: RZAssemblage/RZProxyAssemblage.m [hunk 4]
```diff
@@ -139,9 +120,13 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
             }];
         }
         else if ( changeType == NSKeyValueChangeRemoval ) {
-            [indexes enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
-                [self.changeSet removeAtIndexPath:[NSIndexPath indexPathWithIndex:idx]];
-            }];
+            NSArray *oldValues = change[NSKeyValueChangeOldKey];
+            for (NSUInteger i = 0, index = [indexes firstIndex];
+                 i < indexes.count;
+                 i++, index = [indexes indexGreaterThanIndex:index] ) {
```
Comment (jvisenti) [inline RZAssemblage/RZProxyAssemblage.m:126]: Very against this `for` loop style--it just confused the hell out of me looking at it haha

Comment (KingOfBrian) [inline RZAssemblage/RZProxyAssemblage.m:126]: hah, yea I was pushing it. This is why I try to wait till the next day to merge.

File: RZAssemblage/RZArrayAssemblage.m (hunk 1 above)
Comment (KingOfBrian) [inline RZAssemblage/RZArrayAssemblage.m:80]: Yea, brilliant idea, forgot about those!
Implementing the integer one to directly access, and the object one to access via index path will make the code read very nicely

Comment (jvisenti) [inline RZAssemblage/RZArrayAssemblage.m:80]: Definitely :+1:

**Pull request merged**